### ISSUE TITLE
vagrant / CI: how to handle downloading of packages into VMs at test time?

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#4789]
Taking disucssion from #4766 here.

Problem: we have to download dependencies into the VMs at test time, specifically for the Kubernetes tests, because the version of Kubernetes being tested against is based off of parameters provided at test time. We do this via apt-get:

https://github.com/cilium/cilium/blob/e261cd16b5139ce422cb65a780f32689bd2bbe1d/test/provision/k8s_install.sh#L128-L141

https://github.com/cilium/cilium/blob/e261cd16b5139ce422cb65a780f32689bd2bbe1d/test/provision/helpers.bash#L40-L48

If we want to avoid using apt-get, we'd have to package versions of K8s into the VMs directly. 

However, in the same `helpers.bash` file, I see that instead of using apt, we can download the binary directly:

https://github.com/cilium/cilium/blob/e261cd16b5139ce422cb65a780f32689bd2bbe1d/test/provision/helpers.bash#L50-L69

TL;DR: currently, installing K8s requires apt per the provisioning scripts, but we could possibly avoid using apt all together to avoid EOL issues by acquiring the binaries directly. This would require some exploration, though. We could also package *all* versions of K8s into the VMs directly and install based off of the provided parameter (`K8S_VERSION`) in the CI.

This is an issue because we have various branches which we still are backporting changes to (v1.0, v1.1, and soon v1.2), but we can't backport the VM version changes directly to those branches, because the VM is built against a specific version of Cilium to populate the Bazel cache. @joestringer mentioned that since we are going to be using Ubuntu 18.04 LTS, which won't be EOL'd for 5 years, this won't be an issue for some time, so we could just deal with the pain of building new VMs per release branch for now and not have to deal with it for some time (hopefully we are not backporting to v1.0 5 years from now ;) ).

We can build a version of the VM per release branch. It is possible to build a version of the box with the latest changes to the box (i.e., Ubuntu 18.04 LTS as a base), but have the building of that box check out the release branch and build Envoy / populate the Bazel cache against the version of the release branch by providing an argument to the build pipeline for the cilium branch (e.g., v1.0, v1.1, etc.). Instead of backporting the box, we would open up PRs against the branches themselves directly. The likelihood of merge conflicts is next to none because all we'd ideally be doing is changing a single line in the Vagrantfiles (the box version). 

Packaging the binaries in the VMs is nice because we could avoid having to deal with flakiness when connecting to the outside world, which has been a cause of spurious CI failures.

**Comment by ianvernon** [comment]
cc: @aanm @eloycoto

**Comment by aanm** [comment]
>Packaging the binaries in the VMs is nice because we could avoid having to deal with flakiness when connecting to the outside world, which has been a cause of spurious CI failures.

LGTM. This can be put in the CI image but not in the developer's. We also need to keep the "download" bits from the vagrant provisioning as they are helpful to test different versions when doing a k8s upgrade, e.g. #4757 . Maybe keeping a sort of a cache, does `apt-get` supports having a cache with all packages pre-downloaded?

**Comment by eloycoto** [comment]
Hi,

I'm not sure that I understand the problem here at all, as far as I understand
we have two problems

1) Issues with Bazel dependencies on stable branch
2) Apt repo is out of service.

About the repo out or service, I don't think that move from APT to stable
binaries will fix the problems either, I think that we will point to another
provider, `kubeadm` download all images from Google registry and from Kubernetes
resources urls.

The main issue, if the Docker images are not available is not a explicit error,
we'll see that in the kubeadm timeout kubernetes can't be installed, so it'll be
harder to debug IMHO.

On the other hand, the reason to use `apt-get` is to be as close as the users
use cases, so we can install Kubernetes by hand but I'm not sure that is how the
users are installed it. So, I would rather keep the APT installation for
kubernetes stables branchs and try to be as close as the user as possible.

I understand the problem that we have now with the 1.0 images and the new Ubuntu
18.04 image, the option to be able to update our base image without pain is
important. The only thing that is not working at all is the Bazel dependencies,
that I think that we're not doing well in case of Runtime box.

Runtime is compiling the image in the server using `~/.bazel/cache` dir, and
each time that we update Envoy we need to update the base VM. I would change the
approach here, why we have the dependencies download in our base image? What
about if we compile the base image using docker (As other projects do). Compile
the runtime test using docker container will provide the following:

- VM will not updated so often:
    - Dev specially in the East Europe always complaining about the image size
      and how long it takes ( A few hours for them)
    - `pkg/envoy` team will not need to update the packer-ci-image in each
      build only change the docker image.
- Without ~/.bazel folder, in packer-ci-build we will save around 4GB, that will
  make the builds/download vms faster and we'll pay less money for S3 storage.
- 1.0 Branch will compile the image using container and do the make install in
  the runtime box. So does not matter if what version of Ubuntu we're running,
  the example approach would be like this.

test/docker-compose.yaml
```
version: "2"
services:
  consul:
    image: "docker.io/library/consul:1.1.0"
    hostname: "consul"
    environment:
      - 'CONSUL_LOCAL_CONFIG={"skip_leave_on_interrupt": true, "disable_update_check": true}'
    privileged: true
  etcd:
    image: "quay.io/coreos/etcd:v3.2.17"
    hostname: "etcd"
    command: "etcd -name etcd0 -advertise-client-urls http://0.0.0.0:4002 -listen-client-urls http://0.0.0.0:4002 -initial-cluster-token etcd-cluster-1 -initial-cluster-state new"
    privileged: true
  base_image:
    image: "quay.io/cilium/cilium-builder:2018-06-29"
    volumes:
      - "./../:/go/src/github.com/cilium/cilium/"
    privileged: true
  test:
    extends:
      service: base_image
    depends_on:
      - consul
      - etcd
    command: "bash -c 'cd /go/src/github.com/cilium/cilium/; make tests-ginkgo-real'"
  precheck:
    extends:
      service: base_image
    command: "bash -c 'cd /go/src/github.com/cilium/cilium/; make precheck || exit 1'"
  compile:
    extends:
      service: base_image
    command: "bash -c 'cd /go/src/github.com/cilium/cilium/; make"
```

And to build the image:
```
$ cd $GOPATH/src/github.com/cilium/cilium/
$ docker-compose -f test/docker-compose.yml run --rm compile
$ sudo make install
```

The only backwards here is that the compilation time will take a bit more, if
the builder-image is downloaded will take a few seconds more but the ease use
for all devs, and the benefits will be enough for all. On the other hand, if not
devs need to download a image for a near a hour with the latest dependencies,
etc..

This would be my approach, moving from APT to download binaries will not fix the
issues with Bazel cache that I think that it's the key problem.

Best Regards,
Eloy

**Comment by ianvernon** [comment]
@eloycoto responses inline:

> About the repo out or service, I don't think that move from APT to stable
binaries will fix the problems either, I think that we will point to another
provider, kubeadm download all images from Google registry and from Kubernetes
resources urls.

How does apt-get relate to Docker images? If we are using apt to download the Kubernetes binaries that run on the nodes, e.g., the following which are installed via `apt-get` in the CI currently:

```
14:43:57     k8s1-1.10: Reading state information...
14:43:57     k8s1-1.10: The following NEW packages will be installed:
14:43:57     k8s1-1.10:   kubeadm kubectl kubelet kubernetes-cni
14:43:58     k8s1-1.10: 0 upgraded, 4 newly installed, 0 to remove and 95 not upgraded.
```

These are not Docker images. The question is how we download packages, not images, at test time.

> The main issue, if the Docker images are not available is not a explicit error,
we'll see that in the kubeadm timeout kubernetes can't be installed, so it'll be
harder to debug IMHO.

Have the Docker images ever not been available? AFAIK the images have no specified EOL. Please correct me if I am wrong here. As mentioned before, this is unrelated to the question I posited when filing this issue.

> On the other hand, the reason to use apt-get is to be as close as the users
use cases, so we can install Kubernetes by hand but I'm not sure that is how the
users are installed it. So, I would rather keep the APT installation for
kubernetes stables branchs and try to be as close as the user as possible.

Even if we download the binaries directly / package them within the VM, the majority of the Kubernetes installation would be still done by kubeadm. We would simply package the binaries at some location within the VM for use in the CI; think of it as some sort of cache of dependencies. The code would be very similar to the following, except the binaries would already be in the VM so we wouldn't have to download them:

https://github.com/cilium/cilium/blob/e261cd16b5139ce422cb65a780f32689bd2bbe1d/test/provision/helpers.bash#L50-L69

kubeadm will still do the heavy lifting w.r.t. installing K8s. Thus, I don't understand how using `apt-get` to mirror how users install kubeadm is important here.

The concerns you have about how we build / compile Cilium (building directly on the host) vs. in a Docker container warrant some exploration. I am not against this, but do have multiple concerns:

* Turnaround time for local testing
    - Building Cilium in a Docker image locally for me takes a considerable amount of time when running the K8s VMs - 10-15 minutes, vs. 30 seconds - 1 minute if I've built Cilium in the VM where Cilium is not compiled in a Docker image.
* Building Cilium Docker images within the VMs locally introduces significant load on the host running the VM when compared to direct compilation. 
    - This sometimes even causes kernel panics on the host machines which are running Cilium, especially when building the image for local K8s testing. This has happened for me and others on the Cilium team.

If we can empirically measure the amount of time it takes to build Cilium with the approach that you have proposed, and the team deems it acceptable, I'm all for it.

>  moving from APT to download binaries will not fix the
issues with Bazel cache that I think that it's the key problem.

I am going to phrase what I think you are saying to make sure we are on the same page; you are saying that we would not have to worry about apt issues if bazel was not installed on the host directly, because the cache would be in the Docker image as opposed to in the VM, and we could then easily backport VM changes into prior release branches of Cilium. This means that we would not face EOL issues, and could continue using apt-get like we do now. I think that's valid. As mentioned before, my main concern is how developer productivity would be affected by said changes, as the way Cilium is built heavily affects its development Cilium. This warrants exploration.

One aside - a benefit to not using apt to download binaries is that it causes spurious CI failures because we have issues accessing the repo; if we can remove the need to download stuff from external sources as much as possible, it will remove that cause of spurious failures in the CI.
